Support local parsing of OTPs

thomdixon opened the pull request:
This adds support for the ability to locally parse the OTP and decrypt it (along with some tests). 

This is useful for organizations that program their own OTP slots (either via bulk programming through Yubico or manual programming with the Yubico Personalization Tool/YubiKey Manager), as it will permit them to write their own verification servers in Elixir.

I plan to add more unit tests for hydrating the rest of the fields (by providing the decryption key), and add another option to `:skip_checksum` shortly.

Fair warning: I'm pretty new to Elixir, so feel free to give any and all constructive criticism that you may have.

thomdixon committed:
Add support for local parsing of YubiOTPs

This makes the OTP module a struct representing a YubiOTP. The
OTP can be parsed to hydrate fields on the struct. This permits
servers for validating and rejecting self-programmed tokens to be
written in Elixir.

The exisitng functions on the module have been refactored to make
use of the parsing functionality.

All new functionality has documentation and tests (in the form of
doctests).

binaryape commented:
This is fantastic - thanks!

I was hoping to add some server-side features at some point but didn't expect to get a chance until next year. This is very helpful, cheers!

thomdixon commented:
Sweet! I hope it'll be useful.

I added more doctests showing an actual OTP being parsed, and I added support for the `skip_checksum` option that I mentioned earlier (to decrypt the YubiOTP, ignoring the checksum verification). I also rebased the changes into a single commit, so that if there's any problems, you can easily back them out.

:+1:

binaryape commented:
Sorry about leaving this so long, I've had an odd week. Thanks!
